consulnameresolver/src/main/java/io/datanerds/grpc/ConsulNameResolver.java

```diff
@@ -0,0 +1,123 @@
+
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class ConsulNameResolver extends NameResolver {
+    public static final String DEFAULT_ADDRESS = "localhost";
+    public static final Integer DEFAULT_PORT = 8383;
+    private static final Logger logger = LoggerFactory.getLogger(ConsulNameResolver.class);
+    private final String authority;
+    private final String service;
+    private Listener listener;
+    private Consul consul;
+    private HealthClient healthClient;
+    private Optional<String> datacenter;
+    private Optional<List<String>> tags;
+
+    public ConsulNameResolver(ConsulQueryParameter parameter) {
+        this.authority = parameter.consulAddress.orElse(
+                String.format("%s:%s", DEFAULT_ADDRESS, DEFAULT_PORT)); //todo move to URI
+        this.service = parameter.service;
+        this.datacenter = parameter.datacenter;
+        this.tags = parameter.tags;
+
+        HostAndPort consulHostAndPort = HostAndPort.fromString(this.authority);
+        this.consul = Consul.builder().withHostAndPort(consulHostAndPort).build();
+        this.healthClient = this.consul.healthClient();
+    }
+
+    @Override
+    public String getServiceAuthority() {
+        return this.authority;
+    }
+
+    @Override
+    public void shutdown() {
+        this.consul.destroy();
+    }
+
+    @Override
+    public void start(Listener listener) {
+        Preconditions.checkState(this.listener == null, "already started");
+        this.listener = Preconditions.checkNotNull(listener, "listener");
+        logger.debug("Resolving service '{}' using consul at '%{}'", this.service, this.authority);
+
+        ServiceHealthCache svHealth =
+                ServiceHealthCache.newCache(this.healthClient, this.service, true, buildCatalogOptions(), 5);
+
+        svHealth.addListener((Map<ServiceHealthKey, ServiceHealth> newValues) -> {
+            if (newValues.isEmpty()) {
+                updateListenerWithEmptyResult();
+                return;
+            }
+            listener.onUpdate(generateResolvedServerList(newValues), Attributes.EMPTY);
+        });
+
+        try {
+            svHealth.start();
+            runInitialResolve();
+        } catch (Exception ex) {
+            throw new RuntimeException("Exception while trying to start consul client for name resolution", ex);
+        }
+    }
+
+    private ImmutableCatalogOptions buildCatalogOptions() {
+        ImmutableCatalogOptions.Builder options = ImmutableCatalogOptions.builder();
+
+        if (this.datacenter.isPresent()) {
+            options.datacenter(com.google.common.base.Optional.of(this.datacenter.get()));
+        }
+
+        if (this.tags.isPresent()) {
+            this.tags.get().forEach(options::tag);
+        }
+        return options.build();
+    }
+
+    private void updateListenerWithEmptyResult() {
```

Comment:
Probably rename method to something like `propagateServiceUnavailableToListeners`?